### ISSUE TITLE
PayPal Checkout Refund Failure: "Already Refunded/No Amount to Refund"

### ISSUE THREAD

**Issue by codingwithchris** [woocommerce/woocommerce-gateway-paypal-express-checkout#612]
# Issue Overview
When attempting to process a refund with Paypal Checkout for an order that had not previously been touched, an error occurred stating that the transaction had already been refunded and there was no amount left to refund. Neither of these statements is true in this scenario.

## Issue Production/Reproduction
Here are the steps we took to produce & re-produce this error.

1. Today (August 21st, 2019) we attempted to process a refund in WooCommerce using the "Refund Via Paypal Checkout" button. It was only a partial refund, `$8.01` out of a `$57.73 ` order.

2. We input the refund amount as usual and attempted to process the refund (which correctly read as `8.01` on the "Refund with Paypal Checkout" button )

3. A dialogue came up asking us if we were sure we wanted to continue as this could not be undone. We said "OK".

4. The Processing animation began and after a short while a dialogue box popped up saying:
```
Refund Error: All transactions have been fully refunded. There is no amount left to refund.
```

This is not the case. No refund has ever been processed on this order and the order details confirm this.

5. We logged into our PayPal account and can also confirm that no refund has been manually processed there for this order. To reiterate, there is no record of any refund ever taking place in WooCommerce or PayPal for the order in question.

As an additional note, I believe this issue started after our upgrade to the latest versions of WooCommerce and PayPal Checkout. I am providing additional helpful details as well as a stack trace of the actual error that occurred when the refund threw the error.

We attempted this multiple times from multiple admin accounts and from 2 different office locations (totally different IP addresses) and got the same result.

## Error & System Details
**Error Screenshot**
<img width="2408" alt="Screenshot 2019-08-21 10 16 49" src="https://user-images.githubusercontent.com/17693193/63447959-fa537000-c40a-11e9-9a30-0db48a24ff75.png">

**Here's a stack trace of the error in Sentry**
[https://sentry.io/share/issue/32ec34a1e73e4a46aa0764abe969934e/](https://sentry.io/share/issue/32ec34a1e73e4a46aa0764abe969934e/)

**Brief System Status**
Wordpress: `5.2.2`
WooCommerce: `3.7.0`
Paypal Checkout Gateway: `1.6.17`

**Comment by codingwithchris** [comment]
Actually... to refute my statement about the updates causing the issues... here's a stack trace in Sentry from August 14th that hones in on an unset index `refundable_txns`:

[https://sentry.io/share/issue/e4576d7be09c473f923f86a935fda873/](https://sentry.io/share/issue/e4576d7be09c473f923f86a935fda873/)

So it seems this issue may have existed before the update, I was just not aware of it OR it became worse after the update.

**Comment by fevered** [comment]
Just had another report of this issue in the chat 16270575-hc

**Comment by james-allan** [comment]
Coping in some details from another issue but related to this one.

----


> Payment status is marked as completed, order status set as processing but "_cart_hash" is not created.

That is expected behaviour. Payment is complete/received, the order is set to processing. See [WC Order status explanations ](https://docs.woocommerce.com/document/managing-orders/) for more details.

`_cart_hash` is order meta that is only set if you purchase order through a cart/checkout. Because this is an automatic renewal, there is no cart hash value. 

> Client reported a problem, as for these orders they can't make a refunds.

I don't think any of those reasons would or should prevent merchants from making refunds. To test that I've tried to refund a renewal order and got the following error:

![Screen Shot 2020-01-30 at 1 10 35 pm](https://user-images.githubusercontent.com/8490476/73416866-ede02580-4361-11ea-8431-c22a0450dec4.png)

```
PHP Warning:  Illegal string offset 'refundable_txns' in /app/public/wp-content/plugins/woocommerce-gateway-paypal-express-checkout/includes/abstracts/abstract-wc-gateway-ppec.php on line 371
PHP Warning:  Invalid argument supplied for foreach() in /app/public/wp-content/plugins/woocommerce-gateway-paypal-express-checkout/includes/abstracts/abstract-wc-gateway-ppec.php on line 371
PHP Warning:  Illegal string offset 'refundable_txns' in /app/public/wp-content/plugins/woocommerce-gateway-paypal-express-checkout/includes/abstracts/abstract-wc-gateway-ppec.php on line 395
PHP Notice:  Uninitialized string offset: 0 in /app/public/wp-content/plugins/woocommerce-gateway-paypal-express-checkout/includes/abstracts/abstract-wc-gateway-ppec.php on line 395
PHP Warning:  Invalid argument supplied for foreach() in /app/public/wp-content/plugins/woocommerce-gateway-paypal-express-checkout/includes/abstracts/abstract-wc-gateway-ppec.php on line 395
PHP Warning:  Illegal string offset 'refundable_txns' in /app/public/wp-content/plugins/woocommerce-gateway-paypal-express-checkout/includes/abstracts/abstract-wc-gateway-ppec.php on line 420
```

Looking at the offending lines there is a lot of lines where we're expecting array keys to exist without checking first that the do. I think we just need to fix the `process_refund()` function. 


-----


So after a little more investigation, I think the problem comes from the fact that there are no `refundable_txns` set on the order when it is processed as an automatic renewal. `refundable_txns` is only set in the function `WC_Gateway_PPEC_Checkout_Handler::do_payment()` which is only called via the checkout. 

Processing a payment via `scheduled_subscription_payment()` => `do_reference_transaction()` => `_process_reference_transaction_response()` doesn't set the transaction ID as a refundable one.

**Comment by jrick1229** [comment]
Another report: 19554938-hc

**Comment by chickenn00dle** [comment]
This is resolved in #710 

Closing.
